src/app/services/xml-parsers/structure-xml-parser.service.ts

```diff
@@ -21,63 +21,129 @@ export class StructureXmlParserService {
 
     if (document) {
       const pageElements = document.querySelectorAll(pageTagName);
-      const l = pageElements.length;
-      if (l > 0) {
-        for (let i = 0; i < l; i++) {
-          const element = pageElements[i];
-          let pageContent: XMLElement[] = [];
-          if (i < l - 1) { // TODO: handle last page
-            if (i === 0) {
-              pageContent = getElementsBetweenTreeNode(element.closest('body'), pageElements[i + 1]);
-            } else {
-              pageContent = getElementsBetweenTreeNode(element, pageElements[i + 1]);
-            }
-          } else {
-            pageContent = getElementsAfterTreeNode(element);
-          }
-          // Exclude nodes in <front>
-          pageContent = pageContent.filter(el => {
-            if (el.nodeType === 3) {
-              return !el.parentElement.closest('front') && el.parentElement.tagName !== 'front';
-            }
-            if (el.nodeType === 1) {
-              return !el.closest('front') && el.tagName !== 'front';
-            }
-
-            return false;
-          });
-
-          const page: Page = {
-            id: element.getAttribute('xml:id') || 'page_' + (pagesIndexes.length + 1),
-            label: element.getAttribute('n') || 'Page ' + (pagesIndexes.length + 1),
-            originalContent: pageContent,
-            parsedContent: pageContent.map(child => this.genericParserService.parse(child as XMLElement)),
-          };
-          pages[page.id] = page;
-          pagesIndexes.push(page.id);
-        }
+      const pagesNumber = pageElements.length;
+
+      if (pagesNumber > 0) {
+        this.parseDocumentPages(pageElements, pagesNumber, pages, pagesIndexes);
       } else {
-        // No <pb> used => TODO: Decide how to handle text division
-        console.warn('TODO: Decide how to handle text division when there are no <pb>s');
-        const mainText = document.querySelector('text');
-        const content = Array.from(mainText.childNodes);
-        const page: Page = {
-          id: `page_${new Date().getTime()}`,
-          label: 'Main Text',
-          originalContent: content as XMLElement[],
-          parsedContent: content.map(child => this.genericParserService.parse(child as XMLElement))
-            // tslint:disable-next-line: no-string-literal
-            .filter(c => !!c['content']), // TODO: FIXME: fix property access
-        };
-        pages[page.id] = page;
-        pagesIndexes.push(page.id);
+        this.parseDocument(document, pages, pagesIndexes);
       }
-      console.log(pages);
     }
 
     return {
       pages,
       pagesIndexes,
     };
   }
+
+  parseDocumentPages(pageElements: NodeListOf<Element>, pagesNumber: number, pages: Map<Page>, pagesIndexes: string[]) {
```

Comment:
I'm not sure that passing parsed pages and index maps as parameters is the best choice. I would actually expected to have a `parseDocumentPage` method that given an `XMLElement` returns a `Page` element, so 
that it could be used also for single XMLElements without affecting any parsed data collector variable. 
Something like:
```
parseDocumentPages(pageElement: XMLElement, totalPages: number, pageIndex: number) {
   return { ...PAGE...};
}
```
or
```
parseDocumentPages(pageElement: XMLElement, isFirsPage: boolean, isLastPage: boolean) {
   return { ...PAGE...};
}
```
In any case maybe if you prefer the current implementation, I would change the name of property `pagesNumber`, since it is not immediate that this is the total number of pages (or directly remove it since you can retrieve the total number of page elements directly from `pageElements`). And the name for `pages` and `pagesIndexes` too, to make it more explicit that they are the parsed data collectors.

Reply:
I think this is a better choice: 
`parseDocumentPage(pageElement: XMLElement, totalPages: number, pageIndex: number) {
   return { ...PAGE...};
}`
 but I need also the list of pageElements (or the next page) to retrieve the elements between two <pb> elements.

Reply:
Ok for
```
parseDocumentPage(pageElement: XMLElement, totalPages: number, pageIndex: number) {
   return { ...PAGE...};
}
```
 but I need also the list of `pageElements` (or the next page) to retrieve the elements between two `<pb>`. I'm not sure how to handle this.